File: src/main/java/org/axostudio/axohologram/animation/FrameTextAnimation.java
package org.axostudio.axohologram.animation;

import java.util.List;

public final class FrameTextAnimation implements TextAnimation {

    private final String name;
    private final List<String> frames;
    private final int frameDuration;
    private final boolean loop;

    public FrameTextAnimation(String name, List<String> frames, int frameDuration, boolean loop) {
        this.name = name;
        this.frames = List.copyOf(frames);
        this.frameDuration = Math.max(1, frameDuration);
        this.loop = loop;
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public String type() {
        return "frame-animation";
    }

    @Override
    public long frameIndex(long tick) {
        if (frames.isEmpty()) {
            return 0L;
        }

        long frame = Math.max(0L, tick / frameDuration);
        return loop ? frame % frames.size() : Math.min(frame, frames.size() - 1L);
    }

    @Override
    public String render(String text, long tick) {
        if (frames.isEmpty()) {
            return text;
        }

        return frames.get((int) frameIndex(tick)).replace("{text}", text);
    }
}

File: src/main/kotlin/org/axostudio/axohologram/feature/animation/text/FrameTextAnimation.kt
package org.axostudio.axohologram.feature.animation.text

import org.bukkit.entity.Player

class FrameTextAnimation(
    private val name: String,
    private val intervalTicks: Int = 4,
    val frames: List<String>
) : TextAnimation {

    override fun getName(): String = name
    override fun getSpeed(): Int = intervalTicks

    override fun animate(text: String?, viewer: Player?, tick: Long): String {
        val current = getCurrentFrame(tick)
        return current.replace("{text}", text.orEmpty())
    }

    fun getIntervalTicks(): Int = intervalTicks
    fun getFrameCount(): Int = frames.size

    fun getFrame(index: Int): String {
        if (frames.isEmpty()) return ""
        return frames[index.coerceIn(0, frames.size - 1)]
    }

    override fun getCurrentFrame(tick: Long): String {
        if (frames.isEmpty()) return ""
        val step = if (intervalTicks <= 1) tick else tick / intervalTicks
        val frameIndex = (step % frames.size).toInt()
        return frames[frameIndex]
    }
}
